persistence/test/test.go

```diff
@@ -12,49 +13,51 @@ import (
 
 const PGExpectedSchemaVersion = 1
 
+var dbSuffix uint32
+
 type TestDBSettings struct {
-	DSN                   string
-	Schema                string
 	MigrationsPath        string
 	ExpectedSchemaVersion int64
 }
 
 func PGTestDSN() string {
 	dsn, ok := os.LookupEnv("LNMUX_TEST_DB_DSN")
 	if !ok {
-		dsn = "postgres://bottle:bottle@localhost:45432/bottle_test?sslmode=disable"
+		dsn = "postgres://bottle:bottle@localhost:45432/template1?sslmode=disable"
 	}
 
 	return dsn
 }
 
-func ResetPGTestDB(t *testing.T, settings *TestDBSettings) (*pg.DB, string) {
+func CreatePGTestDB(t *testing.T, settings *TestDBSettings) *pg.Options {
 
-	if settings.DSN == "" {
-		settings.DSN = PGTestDSN()
-	}
-	dsn := settings.DSN
-
-	if settings.Schema == "" {
-		settings.Schema = "lnmux"
-	}
+	// To create a DB, we need a connection to the server that isn't to
+	// the nonexistent DB itself. We connect to the template1 DB, which
+	// always exists.
+	dsn := PGTestDSN()
 
 	if settings.ExpectedSchemaVersion == 0 {
 		settings.ExpectedSchemaVersion = PGExpectedSchemaVersion
 	}
 
-	dbSettings, err := pg.ParseURL(settings.DSN)
-	require.NoError(t, err)
+	dbSettings, err := pg.ParseURL(dsn)
 	require.NoError(t, err)
 
-	db := pg.Connect(dbSettings)
+	defaultDb := pg.Connect(dbSettings)
+	dbSettings = defaultDb.Options()
 
-	_, err = db.Exec(fmt.Sprintf("DROP SCHEMA IF EXISTS %s CASCADE", settings.Schema))
-	require.NoError(t, err)
+	randDbName := fmt.Sprintf("bottle_test_%d",
+		atomic.AddUint32(&dbSuffix, 1))
 
-	_, err = db.Exec(fmt.Sprintf("CREATE SCHEMA %s", settings.Schema))
+	_, err = defaultDb.Exec(fmt.Sprintf("CREATE DATABASE %s", randDbName))
```

Comment:
What happens here if the db already exists because of a failed previous run?

Reply:
Generally, failed runs clean up after themselves (`t.Cleanup()` calls `c.close()` which drops the test DB), but when I created a DB manually and set its owner to the role used by the tests, the code complained that the DB already existed on `CREATE DATABASE`.

I've added a `DROP DATABASE IF EXISTS` above this, so the new code drops the DB and moves on.